AWS-hosted OpenSearch support

Tener opened the pull request:
Add support for AWS-hosted OpenSearch.

Fixes https://github.com/gravitational/teleport/issues/21211.

Tener committed:
AWS-hosted OpenSearch support.

strideynet commented on lib/client/db/dbcmd/dbcmd.go:
```diff
@@ -67,6 +68,8 @@ const (
 	// elasticsearchSQLBin is the Elasticsearch SQL client program name.
 	elasticsearchSQLBin = "elasticsearch-sql-cli"
+	// openSearchCliBin is the OpenSearch CLI client program name.
+	openSearchCliBin = "opensearch-cli"
```
```suggestion
	openSearchCLIBin = "opensearch-cli"
```

Tener replied:
Done here and elsewhere.

Tener commented on lib/client/db/opensearch/profile.go:
```diff
@@ -0,0 +1,100 @@
+	}
+	// calculate config hash
+	h := fnv.New32()
+	_, _ = h.Write(bytes)
```
Not possible in practice.

Tener replied:
Actually, a better guarantee is comment in https://pkg.go.dev/hash#Hash

```golang
	// Write (via the embedded io.Writer interface) adds more data to the running hash.
	// It never returns an error.
	[io](https://pkg.go.dev/io).[Writer](https://pkg.go.dev/io#Writer)
```

Tener commented on go.mod:
```diff
@@ -169,6 +169,8 @@ require (
 	github.com/mailgun/minheap v0.0.0-20170619185613-3dbe6c6bf55f // indirect
 )
 
+require github.com/opensearch-project/opensearch-go/v2 v2.2.0
```
Ah, `go mod tidy` defaults. You are right this looks a bit weird, I'll move it.

Tener committed:
Review: comments, case, cleanup.

smallinsky commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,283 @@
+	defer e.Audit.OnSessionEnd(e.Context, e.sessionCtx)
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := e.process(ctx, req); err != nil {
+			return trace.Wrap(err)
+		}
+	}
```
During testing I'm getting setup on blocking calls: 
```
curl -XGET  'http://localhost:52201/movies/_search' \
   -H 'Content-Type: application/json'  \
   -d '
{
  "query": {
    "match_all": {}
  }
}
'
```
In my case this HTTP call blacks the client forever.

Tener replied:
I'll take a look.

Tener replied:
Running `curl -v` prints a warning: 
`* no chunk, no close, no size. Assume close to signal end`

I chose to calculate the content length (if missing), and now `curl` closes the connection promptly.

smallinsky reviewed (commented):
First pass. Could also add a UX section to PR description.

Tener commented on lib/client/db/dbcmd/dbcmd.go:
```diff
@@ -601,6 +615,36 @@ func (c *CLICommandBuilder) getElasticsearchCommand() (*exec.Cmd, error) {
+		cfg = opensearch.ConfigTLS(c.host, c.port, c.options.caPath, c.profile.DatabaseCertPathForCluster(c.tc.SiteName, c.db.ServiceName), c.profile.KeyPath())
+	}
+
+	// write it
```
Why? This is part of the narrative for the function, where individual comment separate the code into logical parts
```go
// get config
...
// write it
...
// use "teleport" profile and custom config file.
...
// use extra args. if empty, replace with --help
...
```

The "it" here is "config" from the comment above.

smallinsky replied:
There are many Clean Code best practice recommending keeping the comments to minimum  and by replacing obvious comment by a function: 
- https://blog.codinghorror.com/code-tells-you-how-comments-tell-you-why/
- https://blog.cleancoder.com/uncle-bob/2017/02/23/NecessaryComments.html

I know that this is is topic that hooks to the  personal preference but personally I find this rule quite useful. 
For instance we have: 
```go
// getOpenSearchCommand returns a command to connect to OpenSearch.
func (c *CLICommandBuilder) getOpenSearchCommand() (*exec.Cmd, error) {
	// get config
	var cfg opensearch.Config
	if c.options.noTLS {
		cfg = opensearch.ConfigNoTLS(c.host, c.port)
	} else {
		cfg = opensearch.ConfigTLS(c.host, c.port, c.options.caPath, c.profile.DatabaseCertPathForCluster(c.tc.SiteName, c.db.ServiceName), c.profile.KeyPath())
	}

	// write it
	baseDir := path.Join(c.profile.Dir, c.profile.Cluster, c.db.ServiceName)
	tempCfg, err := opensearch.WriteTempConfig(baseDir, cfg)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	// use "teleport" profile and custom config file.
	args := []string{"--profile", opensearch.ProfileName, "--config", tempCfg}

	// use extra args. if empty, replace with --help
	extraArgs := c.options.extraArgs
	if len(c.options.extraArgs) == 0 {
		extraArgs = []string{"curl", "get", "--path", "/"}
		c.options.log.Warnf("No extra arguments provided, using defaults instead: %v.", extraArgs)
	}
	args = append(args, extraArgs...)

	return c.options.exe.Command(openSearchCLIBin, args...), nil}
}

```
but the function can just looks like: 
```go
// getOpenSearchCommand returns a command to connect to OpenSearch.
func (c *CLICommandBuilder) getOpenSearchCommand() (*exec.Cmd, error) {
	config, err := getOpeansearchConfig(c)
        if err != nil {
		return nil, trace.Wrap(err)
	}
       tempCfg, err := writeOpensearchConfig(config)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	// use "teleport" profile and custom config file.
	args := []string{"--profile", opensearch.ProfileName, "--config", tempCfg}
	// use extra args. if empty, replace with --help
	extraArgs := c.options.extraArgs
	if len(c.options.extraArgs) == 0 {
		extraArgs = []string{"curl", "get", "--path", "/"}
		c.options.log.Warnf("No extra arguments provided, using defaults instead: %v.", extraArgs)
	}
	args = append(args, extraArgs...)

	return c.options.exe.Command(openSearchCLIBin, args...), nil}
}
```
or even:
```go
// getOpenSearchCommand returns a command to connect to OpenSearch.
func (c *CLICommandBuilder) getOpenSearchCommand() (*exec.Cmd, error) {
       tempProfile, err := createOpensearchProfile(c)
	if err != nil {
		return nil, trace.Wrap(err)
	}

	// use "teleport" profile and custom config file.
	args := []string{"--profile", opensearch.ProfileName, "--config", tempProfile}
	// use extra args. if empty, replace with --help
	extraArgs := c.options.extraArgs
	if len(c.options.extraArgs) == 0 {
		extraArgs = []string{"curl", "get", "--path", "/"}
		c.options.log.Warnf("No extra arguments provided, using defaults instead: %v.", extraArgs)
	}
	args = append(args, extraArgs...)

	return c.options.exe.Command(openSearchCLIBin, args...), nil}
}
```

Where I don't think that we should always aim to situation where function is as small as possible but 
the comments like: 

> // get config
cfg = opensearch.ConfigNoTLS(c.host, c.port)

for my perspective are considered as unnecessary noise. 

To sum up this is just nit comment and I'm leaving  this up to you.

Tener replied:
I have refactored the code slightly. I believe it is self-explanatory at that point, so I have removed the comments too.

Tener commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3990,6 +3991,50 @@ message ElasticsearchRequest {
+  Metadata Metadata = 1 [
+    (gogoproto.nullable) = false,
+    (gogoproto.embed) = true,
+    (gogoproto.jsontag) = ""
+  ];
+  // User is a common user event metadata.
+  UserMetadata User = 2 [
+    (gogoproto.nullable) = false,
+    (gogoproto.embed) = true,
+    (gogoproto.jsontag) = ""
+  ];
+  // SessionMetadata is a common event session metadata.
+  SessionMetadata Session = 3 [
+    (gogoproto.nullable) = false,
+    (gogoproto.embed) = true,
+    (gogoproto.jsontag) = ""
+  ];
+  // Database contains database related metadata.
+  DatabaseMetadata Database = 4 [
+    (gogoproto.nullable) = false,
+    (gogoproto.embed) = true,
+    (gogoproto.jsontag) = ""
+  ];
+
+  uint32 StatusCode = 5 [(gogoproto.jsontag) = "status_code"];
+  // Path is relative path in the URL.
+  string Path = 6 [(gogoproto.jsontag) = "path"];
+  // RawQuery are the encoded query values.
+  string RawQuery = 7 [(gogoproto.jsontag) = "raw_query"];
+  // Method is the request HTTP method, like GET/POST/DELETE/etc.
+  string Method = 8 [(gogoproto.jsontag) = "method"];
+  // Target is the API target in the X-Amz-Target header.
+  string Target = 9 [(gogoproto.jsontag) = "target"];
+  // Body is the HTTP request json body.
+  // The Struct type is a wrapper around protobuf/types.Struct and is used to marshal the JSON body correctly.
+  google.protobuf.Struct Body = 10 [
+    (gogoproto.jsontag) = "body,omitempty",
+    (gogoproto.casttype) = "Struct"
+  ];
+}
```
I had that thought, but we must be careful with this. There will be custom fields for individual integrations; how do we handle those?

Tener commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,283 @@
+
+	// TODO: ideally we would to switch over to strongly typed opensearch errors, but there is no equivalent of elastic.ErrorCause in opensearch libraries.
+	reason := err.Error()
+	cause := elastic.ErrorCause{
+		Reason: &reason,
+		Type:   "internal_server_error_exception",
+	}
```
Both errors share `type` and `reason` fields. I believe the other fields are optional anyway.

smallinsky replied:
Yes but in Opensearch API response these fields are under `error. type` `error.reason` json path.

Tener commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,283 @@
+
+	roleArn, err := libaws.BuildRoleARN(e.sessionCtx.DatabaseUser, e.sessionCtx.Database.GetAWS().Region, e.sessionCtx.Database.GetAWS().AccountID)
+	if err != nil {
+		return trace.Wrap(err)
```
You can pass account id if you pass full ARN, not just the shortened one. I'll look at how we handle this elsewhere.

Tener replied:
All right, with the recent changes we require the account ID to be assigned at the database level. Whatever you pass as username must agree, but you can use shortened version (e.g. `--db-user=marek-role`) and it will work.

Tener commented on lib/client/db/opensearch/profile.go:
```diff
@@ -0,0 +1,101 @@
+	}
+
+	// write config to file
+	fn := path.Join(configDir, fmt.Sprintf("%x.yml", hash))
```
We can move them, sure. Given content-based names the collision is unlikely anyway. Note that the config files are tiny and removed on `tsh logout`.

Tener replied:
Moved.

Tener commented on lib/client/db/dbcmd/dbcmd.go:
```diff
@@ -601,6 +615,36 @@ func (c *CLICommandBuilder) getElasticsearchCommand() (*exec.Cmd, error) {
 
+// getOpenSearchCommand returns a command to connect to OpenSearch.
+func (c *CLICommandBuilder) getOpenSearchCommand() (*exec.Cmd, error) {
+	// get config
```
This function is tame as far as complexity goes, I don't see the need to extract it, except maybe for testing purposes.

Tener commented on lib/client/db/opensearch/profile.go:
```diff
@@ -0,0 +1,101 @@
+const ProfileName = "teleport"
+
+type Certificate struct {
+	CACert string `json:"cafilepath"`
```
Can you share your `database` config for `os-aws`?

smallinsky replied:
sure: 
```
  databases:
   - name: os-aws
     protocol: opensearch
     uri: search-marek-test-openserach-.....eu-central-1.es.amazonaws.com:443
     static_labels:
       env: dev
     aws:
         region: "eu-central-1"

```

greedy52 replied:
`cafilepath` not needed when connecting to Proxy directly

Tener replied:
After debugging, we were doing _nearly_ the right thing. Indeed the configuration produced for `tsh db config --format=cmd os-aws` will directly call proxy. The `cafilepath` is correctly empty. Were should have excluded that field with `omitempty` though; the updated code does just that.

greedy52 commented on tool/tsh/tsh.go:
```diff
@@ -791,6 +793,7 @@ func Run(ctx context.Context, args []string, opts ...cliOption) error {
 		dbFormatText, dbFormatCommand, dbFormatJSON, dbFormatYAML)).Short('f').EnumVar(&cf.Format, dbFormatText, dbFormatCommand, dbFormatJSON, dbFormatYAML)
 	dbConnect := db.Command("connect", "Connect to a database.")
 	dbConnect.Arg("db", "Database service name to connect to.").StringVar(&cf.DatabaseService)
+	dbConnect.Arg("extra", "Database service name to connect to.").StringsVar(&cf.DatabaseExtraArgs)
```
update description.

Tener replied:
Oops! 🤦 
Thanks, will do.

greedy52 commented on api/types/database.go:
```diff
@@ -423,6 +423,8 @@ func (d *DatabaseV3) getAWSType() (string, bool) {
 		}
 	case DatabaseTypeDynamoDB:
 		return DatabaseTypeDynamoDB, true
+	case DatabaseTypeOpenSearch:
+		return DatabaseTypeOpenSearch, true
```
also when creating a opensearch database, could we throw an error if region is empty, or extract the region from the URI?

Tener replied:
I've updated `RequireAWSIAMRolesAsUsers` and added region extraction and validation to database creation.

greedy52 reviewed (commented):
Tested and worked for me 👍 

One comment is when `tsh db connect --db-user arn:aws:iam::...:role/steve-poweruser  steve-opensearch`, without looking at the source code or debug log, I have no idea what tsh is doing. Is it possible to print some hint that it's curling / and maybe also promote extra args or other tsh options? Or maybe we can disable `tsh db connect` for opensearch (we did this for DynamoDB)

greedy52 commented:
we probably need to update db configurator for AssumeRole permission too.

Tener committed:
Review: comments, region discovery, typos, shared func, fix error type, fix empty CA case.

Tener commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,283 @@
+}
+
+// copyRequest creates a shallow copy of the given request.
+func copyRequest(ctx context.Context, req *http.Request) (*http.Request, error) {
```
I did this for Elasticsearch and OpenSearch. For others, the code is subtly different or uses a different body in cloned requests. We could _probably_ unify these cases, but I'd rather not shoehorn such changes into this PR. This would need to be extensively tested as we can get this subtly wrong.

Tener commented on lib/client/db/dbcmd/dbcmd.go:
```diff
@@ -404,12 +412,18 @@ func (c *CLICommandBuilder) isMongoshBinAvailable() bool {
 }
 
+// isopenSearchCLIBinAvailable returns true if "opensearch-cli" binary is found in the system PATH.
+func (c *CLICommandBuilder) isopenSearchCLIBinAvailable() bool {
```
Fixed.

Tener commented:
> we probably need to update db configurator for AssumeRole permission too.

Yeah, probably a few others too: https://docs.aws.amazon.com/opensearch-service/latest/developerguide/ac.html

A follow-up PR will be suitable to handle that.

smallinsky commented on lib/srv/alpnproxy/common/protocols.go:
```diff
@@ -193,6 +199,7 @@ var DatabaseProtocols = []Protocol{
 	ProtocolSnowflake,
 	ProtocolCassandra,
 	ProtocolElasticsearch,
+	ProtocolOpenSearch,
```
Agree the http protocols can be excluded from `-ping`  support but this `DatabaseProtocols` struct `used for config validation

GavinFrazar commented on lib/utils/http.go:
```diff
@@ -116,3 +116,19 @@ func GetAnyHeader(header http.Header, keys ...string) string {
 	return ""
 }
+
+// CloneRequest makes a clone of the request. Request body is returned as well.
+func CloneRequest(req *http.Request) (*http.Request, []byte, error) {
+	body, err := GetAndReplaceRequestBody(req)
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+
+	clone, err := http.NewRequestWithContext(req.Context(), req.Method, req.URL.String(), io.NopCloser(bytes.NewReader(body)))
+	if err != nil {
+		return nil, nil, trace.Wrap(err)
+	}
+	clone.Header = req.Header.Clone()
+
+	return clone, body, nil
+}
```
I wonder if we shouldn't just use `req.Clone(ctx)` and handle copying the body?

This makes an outgoing request with the same context as the incoming request's context, but it may be better to use the context passed into `HandleConnection`. You could pass that in here though and just use it instead of the request's context.

Since these requests come from `http.ReadRequest(clientConn)`, I think the context is `context.Background()` which means no outgoing request cancellation. I may be wrong though, I just don't see `req.ctx` being set in that method.

Tener replied:
I looked at this; indeed, `req.Clone(ctx)` is much more common in our codebase. At the same time, there are certain assumptions made about request, which make it hard to use the `CloneRequest` correctly. I came to see the `CloneRequest` as unsuitable for shared usage as it is: after the request received from the client is cloned, we need to override `Host` and `RequestURI` fields to be able to issue the request ourselves, among other things. I don't believe singular function is a proper way to share this functionality, but rather we should go in the direction of shared middleware:

https://github.com/gravitational/teleport/blob/a6abd933647c01f311ccfafe8214d3005a250b03/lib/srv/db/opensearch/engine.go#L152-L154

GavinFrazar commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,287 @@
+		return trace.Wrap(err)
+	}
+
+	clt := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+	}
```
may want to use `lib/defaults.Transport` here and overwrite the returned transport's TLSClientConfig

Tener replied:
`lib/defaults.Transport` is preferable indeed, but using it has the side-effect of enabling HTTP2. Typically, this is a good thing, but our code didn't support that use case; we copied the server response as-is, including the mismatched protocol version (we only support HTTP1.1 for client requests).

Disabling HTTP2 would work, but instead, I opted to repackage the response:
https://github.com/gravitational/teleport/blob/a6abd933647c01f311ccfafe8214d3005a250b03/lib/srv/db/opensearch/engine.go#L289-L297

Ideally, we would support HTTP2 throughout the stack, but that needs additional work, which ideally would be shared across the various engines that we have, hence the comment:

https://github.com/gravitational/teleport/blob/a6abd933647c01f311ccfafe8214d3005a250b03/lib/srv/db/opensearch/engine.go#L152-L154

GavinFrazar replied:
good point about HTTP2! I should change dynamodb engine to do the same, since it currently uses the default transport and writes the response obtained as-is.

GavinFrazar reviewed (commented):
First pass, sorry for taking so long to get to this

greedy52 commented on lib/srv/db/common/auth.go:
```diff
@@ -599,6 +599,10 @@ func shouldUseSystemCertPool(sessionCtx *Session) bool {
 		// https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/rds-proxy.howitworks.html#rds-proxy-security.tls
 		return true
+
+	case types.DatabaseTypeOpenSearch:
+		// OpenSearch is commonly hosted on AWS and uses Amazon Root CAs.
+		return true
```
This comment was added before go1.18 mainly as I was afraid that SystemCertPool will not be able to verify on Windows (though i doubt ppl run database agents on windows).

Personally, now I would prefer SystemCertPool for its simplicity so we don't have to worry about tracking what URL to download (and the URL can change too).

Alternatively, we can append this to the system cert pool (IIRC, we do that for azure)

GavinFrazar replied:
Makes sense. I'm in favor of using the system cert pool as a base but also adding the AWS root CA for all our AWS dbs just in case system cert path is overridden by env vars

Tener committed:
Review: address comments.

- add docs
- refactor opensearch engine code
- make `opensearchsql` default client
- drop utils.CloneRequest function, it is not suitable for general use
- drop "extra args" functionality, it should be a separate PR and more generic
- minor refactorings

Tener commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,287 @@
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.Errorf("OpenSearch error: %+v", trace.Unwrap(err))
```
We have this pattern in several places, but you are right. Rewritten to be better.

Tener commented on api/utils/aws/endpoint_test.go:
```diff
@@ -647,3 +647,116 @@ func TestParseDynamoDBEndpoint(t *testing.T) {
+		services []string
+		regions  []string
+		endpoint string
+		wantInfo *OpenSearchEndpointInfo
```
Thanks, removed.

Tener commented on lib/client/db/opensearch/profile.go:
```diff
@@ -0,0 +1,101 @@
+	Certificate *Certificate `json:"certificate,omitempty"`
+	MaxRetry    int          `json:"max_retry,omitempty"`
+	Timeout     int          `json:"timeout,omitempty"`
+}
```
Fixed.

Tener commented:
@greedy52 @smallinsky 

> One comment is when `tsh db connect --db-user arn:aws:iam::...:role/steve-poweruser steve-opensearch`, without looking at the source code or debug log, I have no idea what tsh is doing. Is it possible to print some hint that it's curling / and maybe also promote extra args or other tsh options? Or maybe we can disable `tsh db connect` for opensearch (we did this for DynamoDB)

Marek had a similar comment, and I concur this is not optimal UX. Ultimately `tsh db connect` should run an interactive client, so I replaced the one we had with an interactive `opensearchsql`. The existing client is delegated to alternative command, which is shown by `tsh proxy db`:

```
> tsh proxy db os-aws --port 7111  --tunnel --db-user tener-teleport-role
Started authenticated tunnel for the OpenSearch database "os-aws" in cluster "teleport.example.com" on 127.0.0.1:7111.

Use one of the following commands to connect to the database or to the address above using other database GUI/CLI clients:

  * start interactive session with opensearchsql:

  $ opensearchsql http://localhost:7111

  * run request with opensearch-cli:

  $ opensearch-cli --profile teleport --config /Users/tener/.tsh/teleport.example.com/os-aws/opensearch-cli/2c61534f.yml curl get --path /

  * run request with curl:

  $ curl http://localhost:7111/
```

Tener commented on lib/srv/db/opensearch_test.go:
```diff
@@ -0,0 +1,226 @@
+}
+
+func newTestOpenSearchEngine(ec common.EngineConfig) common.Engine {
+	staticAWSCredentials := func(client.ConfigProvider, time.Time, string, string, string) *credentials.Credentials {
+		return credentials.NewStaticCredentials("AKIDl", "SECRET", "SESSION")
+	}
```
I know, I just didn't want to introduce the dependency for little gain.

Tener committed:
Fix: linting, refactoring issues.

greedy52 commented on api/utils/aws/endpoint.go:
```diff
@@ -731,6 +734,59 @@ func ParseDynamoDBEndpoint(endpoint string) (*DynamoDBEndpointInfo, error) {
+		Partition: partition,
+	}
+
+	// check for recognized service name.
+	if info.Service != OpenSearchServiceName {
+		return nil, trace.BadParameter("invalid OpenSearch endpoint %q, invalid service %q", endpoint, info.Service)
+	}
```
https://docs.aws.amazon.com/general/latest/gr/opensearch-service.html
Seems there are `es-fips` endpionts.

greedy52 commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,331 @@
+		dbRoleMatchers...,
+	)
+
+	if e.sessionCtx.Identity.RouteToDatabase.Username == "" {
+		return trace.BadParameter("database username required for OpenSearch")
+	}
+
```
curious why this check is needed or if it should be checked earlier?

greedy52 replied:
I was just surprised that an empty database user can pass the role matchers check above. Is the intent here to return a bad parameter instead of access denied? Since `err` is returned a few lines above but the `err` is not checked after this if.

Tener committed:
Review: update the comment for accuracy.

Tener commented on api/types/database.go:
```diff
@@ -753,6 +765,40 @@ func (d *DatabaseV3) handleDynamoDBConfig() error {
+	switch {
+	case err != nil:
+		// when region parsing returns an error but the region is set, it's ok because we can just construct the URI using the region,
+		// so we check if the region is configured to see if this is really a configuration error.
```
The code was correct, but the comment was less so; updated.

Tener committed:
Merge branch 'master' into tener/opensearch

Tener committed:
Correct merge issue.

smallinsky commented on api/proto/teleport/legacy/types/events/events.proto:
```diff
@@ -3991,6 +3992,50 @@ message ElasticsearchRequest {
 }
 
+// OpenSearchRequest is emitted when a user executes a OpenSearch request via database-access.
+message OpenSearchRequest {
```
Could you provide aligment for a new message is webapps  repo to cover the new message ?

Tener replied:
The PR for this is coming today.

smallinsky commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,331 @@
+}
+
+// ErrorDetails contains error details.
+type ErrorDetails struct {
```
nit: I think that these internal error struct can be unexported.

Tener replied:
Agreed, done.

smallinsky commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,331 @@
+		return trace.Wrap(err)
+	}
+
+	tr, err := e.getTransport(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
```
nit: in theory the main loop can process many requests:
```go
for {
		req, err := http.ReadRequest(clientConnReader)
		if err != nil {
			return trace.Wrap(err)
		}

		if err := e.process(ctx, req); err != nil {
			return trace.Wrap(err)
		}
	}
```
Can we create transport only once per connection not per requests ?

Tener replied:
Done.

smallinsky commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,331 @@
+		Body:             body,
+	}
+
+	e.Audit.EmitEvent(req.Context(), ev)
```
I think that we should youse internal process context instead of request context that can be canceled aany time by a client leading to interruption of `EmitEvent` flow 
```suggestion
	e.Audit.EmitEvent(e.Context, ev)
```

Tener replied:
Fixed.

smallinsky commented on lib/srv/db/opensearch_test.go:
```diff
@@ -0,0 +1,225 @@
+			dbUser:       "alice",
+		},
+		{
+			desc:         "access denied to specific user/database",
```
```suggestion
			desc:         "access denied to specific user",
```

smallinsky commented on lib/srv/db/opensearch_test.go:
```diff
@@ -0,0 +1,225 @@
+			err:          true,
+		},
+		{
+			desc:         "access allowed to specific user/database",
```
Correct me if I'm wrong but In OpenSearch integration we don't have restrition check for database name so the test name should be: 
```suggestion
			desc:         "access allowed to specific user",
```

Tener replied:
Done.

smallinsky reviewed (approved):
Apart from some nit comment minor suggestion LGTM.

Tener commented on lib/srv/db/opensearch/engine.go:
```diff
@@ -0,0 +1,331 @@
+	e.clientConn = clientConn
+	e.sessionCtx = sessionCtx
+
+	awsSession, err := e.CloudClients.GetAWSSession(sessionCtx.Database.GetAWS().Region)
```
Done, along with minor refactoring to align with the recent changes on the master.

Tener committed:
Review: reuse transport, use different context, rename tests

Tener committed:
Review: unexport internal error types, explicitly test the error serialization.

Tener committed:
Merge branch 'master' into tener/opensearch

Tener commented:
@r0mant please approve this large PR.

r0mant reviewed (approved):
Bot.
